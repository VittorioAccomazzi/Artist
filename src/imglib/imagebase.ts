import { textChangeRangeIsUnchanged } from "typescript";


interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): string {
        return this.pixels[0].constructor.name.replace('Array','')
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}