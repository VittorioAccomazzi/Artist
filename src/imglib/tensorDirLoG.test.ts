import {ImageFactory} from './imagebase'
import CanvasUtils from './canvasUtils'
import { dumpCanvas, dumpImage, getCanvases, hash, overlayTangent, overlayTensor, toCanvas, toNodeCanvas, toSeqCanvas } from './testutils'
import TensorGenerator from './tensorGenerator'
import TensorDirLoG from './tensorDirLoG'
import GaussianFilter from './gaussianFilter'

test('shall detect edges',async ()=>{
    let width = 55
    let height= 88
    let bck = 100
    let frg = 250
    let img = ImageFactory.Uint16(width,height)
    let pixels = img.imagePixels
    pixels.forEach((v,i)=>pixels[i]=bck)
    let xStart = 20
    let xEnd = width-xStart
    let yStart = 15
    let yEnd = height-yStart
    for( let y=yStart; y<yEnd; y++ ){
        for( let x=xStart; x<xEnd; x++ ){
            img.set(x,y,frg)
        }
    }

    let tf = TensorGenerator.Run(img, img, img, 1 )
    let flt= TensorDirLoG.Run(img, tf, 2)

    let canvas = toCanvas(img)
    let tensor = overlayTensor(canvas, tf, 2)
    // await dumpCanvas(tensor,`tensor LoG Test image`)
    let hTensor = await hash(canvas)
    expect(hTensor).toMatchSnapshot()


    let filter = overlayTensor(toCanvas(flt),tf, 2)
    // await dumpCanvas(filter, `shock tensor LoG filtered image`)
    let hFilter = await hash(tensor)
    expect(hFilter).toMatchSnapshot()
})

test('shall find the local image variation',async()=>{
    for await ( const [canvas,ctx,name] of getCanvases() ) {
        let [lImg, aImg, bImg] = CanvasUtils.toLab(toSeqCanvas(canvas))
        
        GaussianFilter.Run(lImg,1)
        let tf = TensorGenerator.Run(lImg, aImg, bImg, 1) 
        let flt= TensorDirLoG.Run(lImg, tf, 1.5)
        // await dumpImage(flt, `tensor LoG image ${name}`)
        let hsh = await hash(flt)
        expect(hsh).toMatchSnapshot()
    }
},60*2*1000)