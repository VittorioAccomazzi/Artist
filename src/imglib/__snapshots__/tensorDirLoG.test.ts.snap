// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall detect edges 1`] = `"be7cf9f3e7cf9f3e7cf9f3e7cf9f3e5cae5cb952a54a952a54a952a54a952a54bf7efdfbe7cf9f3e7cf9f3e7cf9f3e7ca2448952a54a952a54a952a54a952a54ae5cb972e5cb972e5cb952a54a952a54bf7efdfbe7cf9f3e7cf9f3e7cf9f3e7c82040810204081020409122448912244ae5cb972e5cb972e5cb972e5cb972a54bf7efdfbe7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000be7cb972e5cb972e5cb972e5cb972e5cbf7efdfbe7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cb972e5cb972e5cb972e5cbf7efdfbe7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cf9f3e7cf9f3e7cf9f3e5cae5cb952a54a952a54a952a54a952a54bf7efdfbe7cf9f3e7cf9f3e7cf9f3e7ca2448952a54a952a54a952a54a952a54ae5cb972e5cb972e5cb952a54a952a54bf7efdfbe7cfdfbf7ef9f3e7cf9f3e7c820408102043972e5cb9122448912244ae5cb972e5cb9f3e7cf9f2e5cb972a54bf7efdfbe7cfdfbf7efdf3e7cf9f3e7c000000000002952a54a9000000000000be7cb972e5cb9f3e7cf9f2e5cb972e5cbf7efdfbe7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cb9f3e7cf9f2e5cb972e5cbf7efdfbe7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cfdfbf7cf9f3e7cb972e5caa54a952a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7ca2448912244b972e5cb9122448952a54ae5cb972a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c800000000003952a54a9102040810204ae5cb972e5cb9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c000000000002952a54a9000000000000ae5cb972e5cb9f3e7cf9f2e5cb952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be5cb972e5cb9f3e7cf9f2e5cb972e5cbf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cfdfbf7cf9f3e7cb972e5caa54a952a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7ca2448912244b972e5cb9122448952a54ae5cb972a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c800000000003952a54a9102040810204ae5cb972e5cb9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c000000000002952a54a9000000000000ae5cb972e5cb9f3e7cf9f2e5cb952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be5cb972e5cb9f3e7cf9f2e5cb972e5cbf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cfdfbf7cf9f3e7cb972e5caa54a952a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7ca2448912244b972e5cb9122448952a54ae5cb972a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c800000000003952a54a9102040810204ae5cb972e5cb9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c000000000002952a54a9000000000000ae5cb972e5cb9f3e7cf9f2e5cb952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be5cb972e5cb9f3e7cf9f2e5cb972e5cbf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cfdfbf7cf9f3e7cb972e5caa54a952a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7ca2448912244b972e5cb9122448952a54ae5cb972a54b9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c800000000003952a54a9102040810204ae5cb972e5cb9f3e7cf9d2a54a952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c000000000002952a54a9000000000000ae5cb972e5cb9f3e7cf9f2e5cb952a54bf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be5cb972e5cb9f3e7cf9f2e5cb972e5cbf7cf9f3e7cfdfbf7efdf3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cfdfbf7ef9f3e7cf9f3e5cae5cb952a54b9f3e7cf9d2a54a952a54bf7efdf3e7cfdfbf7efdf3e7cf9f3e7ca2448912a54b972e5cb952a54a952a54ae5cb972e5cb9f3e7cf9d2a54a952a54bf7efdf3e7cfdfbf7efdf3e7cf9f3e7c820408102043972e5cb9122448912244ae5cb972e5cb9f3e7cf9f2e5cb972a54bf7efdf3e7cfdfbf7efdf3e7cf9f3e7c000000000002952a54a9000000000000be7cb972e5cb972e5cb972e5cb972e5cbf7efdf3e7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cb972e5cb972e5cb972e5cbf7efdf3e7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cf9f3e7cf9f3e7cf9f3e5cae5cb952a54a952a54a952a54a952a54bf7efdf3e7cf9f3e7cf9f3e7cf9f3e7ca2448912a54a952a54a952a54a952a54ae5cb972e5cb972e5ca952a54a952a54bf7efdf3e7cf9f3e7cf9f3e7cf9f3e7c82040810204081020409122448912244ae5cb972e5cb972e5cb972e5cb972a54bf7efdf3e7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000be7cb972e5cb972e5cb972e5cb972e5cbf7efdf3e7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000be7cf9f3e7cb972e5cb972e5cb972e5cbf7efdf3e7cf9f3e7cf9f3e7cf9f3e7c00000000000000000000000000000000"`;

exports[`shall detect edges 2`] = `"000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000087ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe00000000000000000000007ffffffffe000000000000000000000207339cce60000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall find the local image variation 1`] = `"ffffffffffffffffffffffffffffffffff4f7f3acbfbf5fbfebd3fe5ff977a3fd9f1f1db3a64060ebf4fe639cde91fff8aad7e76e500797fa1f93da2141e9a8983f266915f6f028112072789a1aba013d9190064aafa8f43a530894c5c176acf9ef6fc1f10147fb6f2ecbed547909f21d3bd856027880450dc8ff9afb9672b438e4398bb80537a020168047024c16979f1901da43d7ecdfbd212214832ba1f07ab6b04544e0b02be0fe8d7af65a4d0e18811482ab4dff021000470649b481a2fbaaa535d6b7bfbd68fe4cdbbeea78703f3f51ea15020083b9fafff7e31f7707d811af1d213d4b5c40060ce211a4ccf839a279c256c2a1a3003902086f6d6d16bed54238c20b6e7ef582335fb6bf872e1dba3ff570d6bfb219fe0ab42f986ed2d8ccdad20d6900498829d3641b77fa0d9adfb92af4321cb432020e8ec9ccc132fb2807452a87baff95fc4033d6192e4938c45c2830ffed407ecafc8c2c2ea001db621ec36e58180801e7c9f4c3f20c1ebef9f32ecdb4eb958010380b8d36d6d13c4f44d032cbb64a5ebe8ae830fb0972980c24304551dfa926050c51808ead855bd0c98a8e82100640fa5194f7f0e257fc4a0697b1a4a73bbe8dfe7a1e03df49db7be909da0b410368c209c5698820bf3c851eb1214c6cf8070243cff13993ee5c1be666bfba3250fae3f2fc0f37b832b8c018c000406d84053005f818500300da7fe777082af5f9ff9dfb8476008d76f9d9fbc47f152e060027600853fbfa483b00047a87e7e36fffffb9c258a9a4f3dcb32b056850f822c211ffef3b5e3bd0bacdc1d5170603b53df8033111b6f4ae1d3bb6bad4f9745efe8bfe0cf9008133fbcdb56b2d5059bcc1f61807f7e7f00499f41eaed20f27e3b999007fd23c33433c19c32f09e89598456fdffa986068a5daf3b4103b13bf85bbd964017d9f9cc2bc006fe7ed6c44718216eb7696e6e7f979fd11704ddd52b97f283ddffb75dc65fe484e0fa033bf4080be420005d84d0639f721bc5fe4503951839dfffae0fbe9ff5fdf61310eeecfbe7d73a6ccbf872e3e0116bfe5e939ebead7c38011435a9dbf2e80163f5e67d83fa07e09a75db633f9ea4febc1a10466cd48e8841a20c1ba7c9fb81dfbfe3f9bb6fed97f34cfca097e2d20e60cc6d464183128e8fe1bbdb13cb14f3dd49fafdfff5687bf01e11bdefa57b68271c0fff88e94462074508469be40cf2cd73f840977fbebf7ab9457a45affb9fb1a506836011c1f8c7cdf64a31d600300e3af0ff7ff0284d3873c7a001fcd1c9fddc1fbfa40fb2c3b7a400000183ecffea037fc1096000400010040bfff3376c59ffe6b7fffff0080002fff60f842e979f2811fc010001bfffff030273b9ff836003640800000ffe8df3b87799cd557c97f83000fffffa50a8edb74c63c20a20280e1727ff95ed4d9a4b91fd9fbffbbd2ff000cc00603187affe9600fdff1ece500e0f7f9f8b8fac7e0001ff0180000000000bffe1f2b07000840000018000000001fe001a75ff80001fffffff80001fffffffffffffffffffffff021fffffffe0000001ff1beb9cc8080c7001ff80000181fffe80000000007ff3fffff88afffe3fef787fffff000000000001c8e5004888139fe04000b8e944d27fbf8f2e2337ff880010137fdffab9ffc417e050c00000e76055fff9a803140011e9ac0adffeff3c5be208ec1738cffeee31fffd77bf80820e081f07d8c33a02115f8001e3f55700d1f520400002418180218ffe0c001fff0f028fbe0c18823c1891fc84001ccec62c7c7c03f7e7feffffff83f0dfb9211fd80fa7f01c003be1b711f80000c7fe60e3f4101f03fff73e78279fffffb003df103fef81fff1fbc1dff9e001cf481cc3ffe07078000001ff13fdfffe00ffc03c0003661ffffff001e03f9c01ff0ffe03fffc00de00340e341e4183fc03f000ff0007ff00dcc1fdfffffffdfff0001fe1dffc007fffe700038001f0000fffe03ff0003fd080003cfc7fff9e3febc19fc407cfffce7bffe3a7c001e7e0003e7040f8f0e03084ec3f580ffff01f8fd8003db7800fc13f930020f00f8021f887fffff9bfc0fff11fffefc3e1fff8003ee00000007f00006000107ff5c207ffc1000efbffc1ffeffffeff401f9c00000004fbcfe01f03f9f01f87bc43a3bfffc3808000000010200fe058c7ffffc0003c7fffffffffffff0001304fe3800f8787f803f80100038fffefc7f801fee27fc001e0001fbf7c0000387b02efa7f9803fffffffe04001fe00063c7ffb800e7f9980dfbf17dffffffff0838019c23180601e80400c000000000f227f01fffe671cc07ffffffffbf87fd1dc01cf810180017f800008006efffffc33f83ff8fc7ffff0706623f983800003cc6601830780018fffffff1e7c0fe1e00189fffc23ff3f88000000639ffefe3f3fff9f87e000f0607ffff99fc00031eff00001901fdc07ffffb81f0003ffe050001ffdefe1fbe040000183ffff901f81e00081b3180c3fb800e67c3f3867800fdffe1fddeef38027ffff8300f1805df0300001a0419effecf8437fff847fffffe7e7ffc31f4000000038007c1f8020063fffff80003effff7f7f3800e07fe67843a0c1dffcf196c03a81c7efbdc00d8008000ff8679fe07ffc383dfc0f18ffae8087e187dbe07f8e03ffc783f00200000219ff8c7c3e00e01c21a03c0fc0e03fffe07fe303d0c60fe7ff3fe37839ff8f0a019988300ffd7f7bf801c7cfdf01f3fffc03b7ffd78fe7dc01fe3030209e0c1cf27fdf70386fba00c6030f800d3027e21e8f8003831800ef3cfde73ff201e1bfc2e1fe7e6483c4faefffdbc0fdff7c043f43fdc79c0e01fc00e7e01e0667dff01dffffffffffffffffffffffffffffffffff"`;

exports[`shall find the local image variation 2`] = `"00000000000000000000000000000000000000000000000000000000000000001b81d5e5dfbd7fefbfe7c3fd7c3977e03f01f1e8f6efeef2d6aac19d6c1d3dc43f11dad62b14a1b2932ae1a93e0c078c3770c7553fafa63bd7ad60d76b0cfb181611c9b76bef77febd6df17d7b071e3c3c31c5675f073ea7bdc6e0dbdb86bc783c31d3c81a4a7f5d5571e96d618358d01870c77ea7bb7efb37e261f526c101c031b1edf75fbd6ff85aef62afabc1e3dc27b1c3d59a577e1f8feb739ef660c71021b0ced6f5a78801f14573c84fe0cf4c0331c5e3267a39387bab63c5f7705a500310cb62fdc60ffe1caef167fff83c140630c57db39ff5530ea9f3b39a987d6c0630eff77ef4fd65e303e1f3df9c7acc0e20d397cd849c3ec9c77081dfbcf5980c10c7fb75752f7930e37080dfa9a3301c00e7e35bda95871c327020dff3e6743e20dffe33a7fbdce198f21879e7c4ec3600d2e767478ad38a8c710c6ae7ad982610c8da6eadf5c633cc608c2ccf1bf02e10c7abd93907988e67610c179eb7e02610c733d67fe0b2f7f3a08e1e9c25c41620c561b3d77dcebc30e18f0b3d698c2e31cdedbbf0ab8fc554e08f8e785b982600d6f57e19d6332a7a620dbf7393301e10ccb3791c9c8c61ca630dc0d6a6643630d7f37b7330d5ae23220f98e36ccc0610eaf26adc626538ff328f20e2dc982610e172773ac5b26aa19a0cce7a99bc1601c7b66c719a4d851d4c191c4fb33c0621c2d6f2cb5533b122c5e27c41327c3620d1e4e69e58d69cc477ccf8d366d43601d464ff69a5cb7bb9b333ccd7e4cc1621d3e4f8594634c64348c794e5c9fc3e00c7ec9932846b59aeb39f84fd99e41e10d2cdfa611c98a6936c3631db91782e01c5adc69267251a7998e669ab32783e11cfcdad2df2c6e52e63ee59b723fc3e11c4acae586d318cdd8e8c8b3667741e00c26caeb4d80e00333f3c1b2e4f7c3e20c7cca565b7f3812cc0b8736cd7f83e00c2cd27c764f07f99d243e64d935c3e10e2cd2e854b1baf23270fced93aec1f20ca8c371990786ec57c3f1dd93fbc3e10ce94233b2372ad35f8fc39926d482e30cbe6b6262de66a63e38072b291503e10cea6244d9b4ca4cfc6c4ed26dedc3e30cbe22890f2dae9ae0e49c16427d41e20cc630d882edd31d9cd639768f6dc2e10cfb18f338ca1247f86a7a6c8cfb43610e3718ac1de6ee9cca68f4ed8c77c1600c5719986d56d03b8e74f6d90f9a43600eeb99f05e51a476287cfcd960de41600f7f7967535724ceb83ccd9a7e0bc2610e1eed83d74a445aaa3cf5b27f8082e00c2e8d86d54aa4f3623ecdb61ff881610d0f7b0ae6e964f53d2ef364e2df03600c57908ace9a6103923edb64007fc2600d3e6716fa9467c3983e5769fd8fc3620e37e6748a744c7990ee4549ffc1c1610e4fd26ab72ad01c930e76caef5040600e0791cab6d43fe683e666d19dd903600dbe470ab9c47ff78fee6e93e99282e10cf656739fa8f7258f6f6d95d7b6c0610c461cc35bd9b73f967d5da5f447c1610c5c398b6b039815987d5d22da0f81610c506771f6a607cb93dd5f2df9fe01600c4371d9be8c332b9c9d7b4bba9c03600c6ec5ada519401b9a197a47e430c2610dc1cb93bd13519f96193a5d2ae142600f0fa7c4fa3344339b9d3e9abdc2c2600edb4bf4fa6361139b9d3697e9cfc3618fe6ee747471791bcd9934b5f18f02610d19e4fd34f19b1bc9313436fb9e02610c7744e8a9d19d9bcd391d128b1c02600fba6e53e9d0df1bcd391d7dbe38c0610cb34a6bdbd1aa10cd79995e2673c0210b32437bc3d16e9348719851b667c0200c36d3f787d2cc19ccf19a68cc6682e30d2d53a66dd7dd0398f39aadeccec0e30d2949b8dbd1990698a692a82ccfc2e10f3769709ed19cae99e590fb8c9d43e28ea4a879bed1de3299f2942f1d3740e10ca5b6c36ed80c011b6294f50c2602e00d2d89a6ec58ebfff352d4b6987f02e11b7ca5a4a548906846d2d5fd985f00a1866db229b548709046b2d56318fc03a00c49092bbf491c7fceeadcbad8e9c3a10d5b6ce6a94cbc7f1f6ad90198e7c3a88ddb68e7ac6e13e01e6ac89e98ce01a98c9b466daa2a06e1aeaac81598f003a98cdb627dab7582c4aeaa48f198c303a80ddb357dabb5e04026ea481b99c003a10cb3330d2dd4fffffae2ce00191043a192971d272dd600003eeacfe031a1c3a126db89b32ca69f7f076adbfc1383c331cddd93a12eef2dfe01e2d0fff38701f089eccab5aa7306fe60e6d51fb28c80f113c1974aeb7b2c7a3236d701b2198070b7d32aaaef6b1b73d5149669b0330031b31fecd56f6b32f3f48c9731b0324031268dd5466f6b49fb7e4caac7a46e403012ad52b267711c6bf766aee1e46cc031972d6667677116cfdd67aaab80d1803197296e63677511fb7573a56720f3803196f92f6425b116dfd6339daf21e3003105b1b502e4e10f75db799d2e21e7003187a9b58c3ae312fef3490f2e63e6423006b9458782e31eeb54292f8e63e6c23005e89552db8319deddb58bde226c8230979b5a89a7821ffadd5480ff86d9c231973b1aeceb8602cff648a019c69b823097f2d67eafa6567d5da2fc0985b30210876a697bdf24457f96b45f838f6702108aca27075e0ccdfd56ae4ff58eca0210978adb9abe18f9773dae61f98ece4318967d3dcafa30a77b4f9760791d9cc31897eef6f73c633d5d6b2722183b3c8318976e9acb71c3fffbb9d53148763943189636da6ebf0a5a35def1314dec70c3188f96cb477c06d7affb5010cfd8e0c3188e96cd26e00e35ebfff69897b961c3189faacdb7c159db4b7eaa9b3333c180000000000000000000000000000000000000000000000000000000000000000"`;
